### ISSUE TITLE
[Question] Benefiting of having multiple CPU cores

### ISSUE THREAD

**Issue by valeriyvan** [rism-digital/verovio#1490]
At the moment Verovio is using single CPU core for everything: parsing xml, laying out document and rendering pages.

Even slowest mobile devices have now multiple cores. Newest iPad Pro has as many as 8. And Verovio is using only single core.

Is there any way of benefiting from having multiple cores?

**Comment by lpugin** [comment]
I am not sure how this could work. I think this would come with quite significant additional complexity, but I am not and expert. If somebody has ideas, please comment here - closing this issue since there is not action to be taken.

**Comment by valeriyvan** [comment]
May be it does make sense to keep this issue open for some time. Otherwise no-one will see it and therefore has no chance to share ideas. Issue is labeled as Question, so it's clear it will be no action on it.

**Comment by lpugin** [comment]
OK, let's give it a month

**Comment by craigsapp** [comment]
A general problem is that javascript is a single-thread (core) system, so any multi-core implementation would most likely be incompatible with the javascript toolkit version of verovio.  So there would probably need to be a separate Toolkit class that works with multiple threads for use in environments such as a Qt interface.  There are web workers to split computationally intensive tasks into separate threads in javascript, but that is only useful to separate the user interface from the verovio toolkit, not to split processes within the verovio toolkit.

> At the moment Verovio is using single CPU core for everything: parsing xml, laying out document and rendering pages.

For a multi-core implementation, the computation would need to be possible to do in parallel.  A data file first needs to be parsed, then layed out and then rendered, which does not particularly fit well into a parallelized model.  The rendering process is the most parallel like: once a decision is made about how to allocate data into separate pages, each page could be implemented in a separate thread.

**Comment by valeriyvan** [comment]
> A general problem is that javascript is a single-thread (core) system, so any multi-core implementation would most likely be incompatible with the javascript toolkit version of verovio. 

Verovio is C++ library, not Javascript. One approach would be creating multithread aware toolkit for appropriate environments. Another approach would be modifying existing toolkit, e.g. adding parameters for different threads for different workers. For single thread environments like Javascript these will be same thread, when multithread environment could supply different threads for different workers.

**Comment by valeriyvan** [comment]
> > At the moment Verovio is using single CPU core for everything: parsing xml, laying out document and rendering pages.
> 
> For a multi-core implementation, the computation would need to be possible to do in parallel. A data file first needs to be parsed, then layed out and then rendered, which does not particularly fit well into a parallelized model. The rendering process is the most parallel like: once a decision is made about how to allocate data into separate pages, each page could be implemented in a separate thread.

Parsing document and laying out document could be done in parallel in separate threads. There's no need to have whole file to be parsed to start laying out. Laying out processes doesn't consume all data as one chunk. It traverses tree. Parsing could do something like that and feed new pieces of data to layout worker as it parses file.

**Comment by lpugin** [comment]
I guess it depends on how you want to use it. You do need the data to be loaded to perform the layout, which is why this is sequential. Of course, if you want to load and render some parts independently, the you can split the data upfront and launch distinct loading / rendering processes. As you pointed out, Verovio is a libary, so it is no problem of make a multi-threading application around it that create as many instances of the toolkit you need - e.g., one per core will be optimal.

**Comment by craigsapp** [comment]
Pugixml is thread-safe: https://pugixml.org/docs/manual.html#dom.thread

Here are benchmarks for pugixml: https://pugixml.org/benchmark.html

I doubt that the XML data can be loaded into pugixml at the same time that other processes are reading from it.  For example, I cannot find a callback feature in the pugixml documentation when it is done parsing an element (but I did not check for very long).

However, verovio has its own separate structure for the MEI document, loading in from the pugixml data structure, so the pugixml-to-verovio converter could send signals to a separate thread when another measure has been loaded, allowing the layout process to start before the verovio MEI data structure has been fully converted from the pugixml structure.

**Comment by valeriyvan** [comment]
> Out of curiosity: what piece do have that is 50MB? 

L. van Beethoven, Op.125 NEUNTE SYMPHONIE
Mei file is 61Mb.

>And when you have a symphony, do you have one file per movement? Or one file for the entire symphony? It would be best to split them by movement.

That's obvious solution. But. We would like to have a feature to render movement not from new page but continuing page of previous movement. Some compositions have not closed movements (when music doesn't interrupts between movements) in which case continuing new movement on current page is the only correct way of rendering (https://github.com/rism-ch/verovio/issues/1262). The only way to handle these cases - process all movements in one run. It seems like probably we can't avoid situations where we have load whole Mei file with all `mdiv`s.

> It should not be that big. I heard about people working with a Mahler symphony that was 10MB. The Hummel concerto in the examples is 6MB for all the three movements.

May be extreme cases we have are malformed as they were converted from MusicXML.
> From my experience, loading an XML (MusicXML) file that is that big in a music notation editor also takes quite a few seconds. There are certainly ways to make Verovio faster, but I doubt you can easily make it, let's say, 10 times faster, which as I understand would be the minimum for you to be acceptable.

We should try. Not utilised CPU power seems like opportunity.

> Whatever you achieve, I think it will remain difficult to do instant responsive layout with 50Mb or more of data. On this topic: when you say "changing settings". What are these? Page / screen dimension? Zoom? or other settings? Maybe we can look at ways not to redo the full layout when this might not be necessary. But if you change the page dimensions, then you do need to do quite some calculation to get it right - and if you are currently looking at the end of the piece, then I cannot see how to calculate the layout without starting from the beginning of the piece.

Settings we are working on:
* change of musical font
* change of musical font size
* change of line spacing
* change of line spacing
* change of page margins 

Change of any of these setting supposes immediate feedback. User should see re-rendered page as soon as possible.
 
> At the end of the day, it is a question of priority. For the Verovio project in general, I think that improving engraving quality has still by far a higher priority than increasing speed. Rendering music notation is a quite complex process, and we should think twice before adding the complexity of multi-threading directly in this process.

Understood. 
Hope we could find ways to make Verovio faster as well.

**Comment by wergo** [comment]
Just to add the case of Mahler IV which we converted within TROMPA: 
1st mvt: 10.6 MB
2nd mvt: 7.9 MB
3rd mvt: 7.6 MB
4th mvt: 5.1 MB.

In MEI-friend (https://atom.io/packages/mei-friend), we have tried to tackle this problem by just parsing the requested page/system to Verovio which gives great speed improvements. Of course, you will need a page/system layout first (which we almost always have in our encodings).
